### DISCUSSION TITLE
Confusion about schema error messaging

### DISCUSSION THREAD

**Discussion by zaydek** [open-circle/valibot#634] [Q&A]
I think this relates to: https://github.com/fabian-hiller/valibot/discussions/467.

I'm a bit confused about how to take advantage of custom error messages. For example:

```tsx
import * as v from "valibot";

const StringRecordSchema = v.record(v.string(), v.string(), "An object is required.");

const data = {
  foo: 5,
};

function main() {
  v.parse(StringRecordSchema, data);
}
main();
```

I would expect the stack trace would consist of my custom error message but alas I see:

```
zaydek@Zaydeks-MacBook-Pro orderediting-jobs % bun run analytics/schemas/statistics.ts
190 |    * Creates a Valibot error with useful information.
191 |    *
192 |    * @param issues The error issues.
193 |    */
194 |   constructor(issues) {
195 |     super(issues[0].message);
          ^
ValiError: Invalid type: Expected string but received 5
      at new ValiError (/Users/zaydek/Code/__newschool/orderediting-jobs/node_modules/valibot/dist/index.js:195:5)
      at parse (/Users/zaydek/Code/__newschool/orderediting-jobs/node_modules/valibot/dist/index.js:4831:11)
      at main (/Users/zaydek/Code/__newschool/orderediting-jobs/analytics/schemas/statistics.ts:52:3)
      at /Users/zaydek/Code/__newschool/orderediting-jobs/analytics/schemas/statistics.ts:54:1

Bun v1.1.12 (macOS arm64)
```

I'm also aware there's `safeParse` but when I inspect the issues and or the first issue's path I still don't see the error message. Am I misunderstanding how this API works? I am basing my example off of the docs for records here https://valibot.dev/api/record/#generics.

Any help is appreciated, thanks!

**Comment by fabian-hiller** [comment]
Any schema has its own error message:

```ts
const RecordSchema = v.record(v.string('key error'), v.string('value error'), 'object error');
```

If you want the same error message for all nested schemas you can use the `config` method or write your own [`message` method](https://github.com/fabian-hiller/valibot/discussions/463#discussioncomment-9694410):

```ts
const RecordSchema = v.config(v.record(v.string(), v.string()), { message: 'error' });
```

If you want the same error message for all schemas when using `parse` or `safeParse` you can use the `config` argument:

```ts
const RecordSchema = v.record(v.string(), v.string());

const output = v.parse(RecordSchema, data, {  message: 'error'  });
```

If you want the same error message for all schemas globally you can use `setGlobalMessage`:

```ts
v.setGlobalMessage('error');
```

**Reply by zaydek** [reply]
Interesting. I'll give this a try and report back. Thanks.
